pkg/util/token_bucket.go

```diff
@@ -0,0 +1,95 @@
+package util
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type TokenBucket struct {
+	remainingTokens int64
+	maxCapacity     int64
+	refillRate      int64
+	lastRefill      time.Time
+	mu              sync.Mutex
+
+	remainingTokensMetric prometheus.Gauge
+}
+
+func NewTokenBucket(maxCapacity, refillRate int64, remainingTokensMetric prometheus.Gauge) *TokenBucket {
+	if remainingTokensMetric != nil {
+		remainingTokensMetric.Set(float64(maxCapacity))
+	}
+
+	return &TokenBucket{
+		remainingTokens:       maxCapacity,
+		maxCapacity:           maxCapacity,
+		refillRate:            refillRate,
+		lastRefill:            time.Now(),
+		remainingTokensMetric: remainingTokensMetric,
+	}
+}
+
+func (t *TokenBucket) Retrieve(amount int64) bool {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	t.updateTokens()
+
+	if t.remainingTokens < amount {
+		return false
+	}
+
+	t.remainingTokens -= amount
+	if t.remainingTokensMetric != nil {
+		t.remainingTokensMetric.Set(float64(t.remainingTokens))
+	}
+
+	return true
+}
+
+func (t *TokenBucket) ForceRetrieve(amount int64) {
```

Comment:
Does it make sense to combine `ForceRetrieve` and `Retrieve` and make `force` a parameter?

Reply:
I separated them because:
1. I wanted `Retrieve` params to be as simple as possible, since this is the most commonly used function for token bucket
2. When we are force retrieving, the function doesn't have to return anything

Reply:
Make sense.